### ISSUE TITLE
[java] CloseResource possible false positive with Streams

### ISSUE THREAD

**Issue by oowekyala** [pmd/pmd#1922]
**Affects PMD Version:** 6.16
**Rule:** [CloseResource](https://pmd.github.io/latest/pmd_rules_java_errorprone.html#closeresource)
**Description:**

CloseResource reports that Streams should be closed (indeed they implement AutoCloseable). In nearly all cases those violations are noise though. Closing a stream that doesn't feed on system resources is useless. Only streams that are obtained from eg [Files::lines](https://docs.oracle.com/javase/9/docs/api/java/nio/file/Files.html#lines-java.nio.file.Path-), or [Files::walk](https://docs.oracle.com/javase/9/docs/api/java/nio/file/Files.html#walk-java.nio.file.Path-int-java.nio.file.FileVisitOption...-) should really be closed, but they could escape their defining scope and be closed elsewhere (the laziness of the stream API is meant for that). I don't think we can reliably flag those at the moment. So I would say that the rule should not by default consider Stream instances.

**Code Sample demonstrating the issue:**
```java
 public static <T> Stream<MatchResult<T>> filterResults(List<T> candidates, Function<T, String> matchExtractor, String query, MatchSelector<T> limiter) {
        if (query.length() < MIN_QUERY_LENGTH) {
            return Stream.empty();
        }

       // violation here
        Stream<MatchResult<T>> base = candidates.stream()
                                                .map(it -> {
                                                    String cand = matchExtractor.apply(it);
                                                    return new MatchResult<>(0, it, cand, query, new TextFlow(makeNormalText(cand)));
                                                });
        return limiter.selectBest(base);
    }
```

**Comment by adangel** [comment]
Agreed. It's explicitly in javadoc: https://docs.oracle.com/en/java/javase/12/docs/api/java.base/java/util/stream/Stream.html

> Most stream instances do not actually need to be closed after use, as they are backed by collections, arrays, or generating functions, which require no special resource management. Generally, only streams whose source is an IO channel, such as those returned by Files.lines(Path), will require closing. If a stream does require closing, it must be opened as a resource within a try-with-resources statement or similar control structure to ensure that it is closed promptly after its operations have completed. 

I'd suggest to add java.util.stream.Stream to the ignore list by default.
